src/tunnel.c

```diff
@@ -0,0 +1,276 @@
+            if (nread <= 0) {
+                ENSURE_CONNECTED_OR_RETURN(pipe->read_conn, session);
+                break;
+            }
+            pipe->buffer_len = nread;
+        }
+        if (pipe->buffer_pos != pipe->buffer_len) {
+            pipe->buffer_pos += connWrite(pipe->write_conn, pipe->buffer + pipe->buffer_pos, pipe->buffer_len - pipe->buffer_pos);
+            ENSURE_CONNECTED_OR_RETURN(pipe->write_conn, session);
+            if (pipe->buffer_pos != pipe->buffer_len) break;
+        }
+        pipe->buffer_len = 0;
+        pipe->buffer_pos = 0;
+    } while (1);
+
+    if (pipe->buffer_len == 0) {
+        connSetWriteHandler(pipe->write_conn, NULL);
+        connSetReadHandler(pipe->read_conn, callback);
+        return;
+    }
+    connSetWriteHandler(pipe->write_conn, callback);
+    connSetReadHandler(pipe->read_conn, NULL);
+}
+
+static void upPipeTransfer(connection *conn) {
+    tunnelSession *session = ((client *)connGetPrivateData(conn))->tunnel_session;
+    pipeTransfer(session, &session->up_pipe, upPipeTransfer);
+}
+
+static void downPipeTransfer(connection *conn) {
+    tunnelSession *session = ((client *)connGetPrivateData(conn))->tunnel_session;
+    pipeTransfer(session, &session->down_pipe, downPipeTransfer);
+}
+
+static void initTunnelPipe(tunnelPipe *pipe, connection *read_conn, connection *write_conn) {
+    pipe->read_conn = read_conn;
+    pipe->write_conn = write_conn;
+    pipe->buffer_pos = 0;
+    pipe->buffer_len = 0;
+}
+
+static ConnectionType *connTypeOfTunneling(void) {
+    if (server.tls_replication) {
+        return connectionTypeTls();
+    }
+    return connectionTypeTcp();
+}
+
+static const char ok_str[] = "+OK\r\n";
+static const size_t ok_str_len = sizeof(ok_str) - 1; // 5
+static const char queued_str[] = "+QUEUED\r\n";
+static const size_t queued_str_len = sizeof(queued_str) - 1; // 9
+
+static int isReplyError(const char *buf, size_t buf_len) {
+    return (buf_len > 0 && buf[0] == '-');
+}
+
+static int isEntireReplyReceived(size_t buf_len, size_t len) {
+    return buf_len >= len;
+}
+
+static int isExpectedReply(const char *buf, const char *str, size_t len) {
+    return (memcmp(buf, str, len) == 0);
+}
+
+static size_t expectedReplyLength(int multi_cnt) {
+    size_t ret = ok_str_len + multi_cnt * queued_str_len;
+    return multi_cnt ? ret + ok_str_len : ret;
+}
+
+static int verifyReadReply(tunnelSession *session, const char *buf, size_t buf_offset, size_t buf_len, const char *expected_str, size_t expected_str_len) {
+    if (buf_offset >= buf_len) return 0;
+    buf += buf_offset;
+    buf_len -= buf_offset;
+    if (isReplyError(buf, buf_len)) {
+        abortTunnel(session);
+        return 0;
+    }
+    if (!isEntireReplyReceived(buf_len, expected_str_len)) return 0;
+
+    if (!isExpectedReply(buf, expected_str, expected_str_len)) {
+        abortTunnel(session);
+        return 0;
+    }
```

Comment:
Supposing it returns a response that isn't an error but is shorter than expected, would the connection just be left hanging?

Instead of checking error followed by checking if it is expected, I think this may be simpler/less error prone:


```suggestion
    size_t prefix_size = buf_len > expected_str_len ? expected_str_len : buf_len;
    /* Check if what is in buf partially or fully matches expected_str */ 
    if (strncmp(buf, expected_str, prefix_size)) {
        /* What we have so far does not match what we expected */
        abortTunnel(session);
        return 0;
    }
    if (!isEntireReplyReceived(buf_len, expected_str_len)) return 0;
```